### PR TITLE
[contextmanager-generator-missing-cleanup][new feature] Warn about @contextlib.contextmanager without try/finally in generator functions

### LINKED ISSUE #2832: Warn about @contextlib.contextmanager without try/finally in generator functions
### Is your feature request related to a problem? 

If you decorate a generator function with `@contextlib.contextmanager` and use the context manager `cm` inside a generator function, the context manager's `__exit__` will not be always be called because the generator raises `GeneratorExit`, which ends the generator that defines `cm`. 

Notice that `cm exit` never appears, as `cm`'s `__exit__` is never called.

```python
from __future__ import generator_stop

import contextlib
import weakref
import sys


@contextlib.contextmanager
def cm():
    print("cm enter")
    a = yield
    print('cm exit')

def genfunc():
    with cm():
        print("stepping")
        yield

def main():
    gen = genfunc()
    ref = weakref.ref(gen, print)
    next(gen)


if __name__ == "__main__":
    main()

```
Output:

```
cm enter
stepping
<weakref at 0x7f4739e38e08; dead>
```

### Describe the solution you'd like

Generator functions decorated with `@contextlib.contextmanager` should not be used in generators unless the context manager has `finally:`

```python
@contextlib.contextmanager
def cm():
    try:
        yield
    finally:
        # cleanup code here
        ...
```

or specifically handles `GeneratorExit`:

```python
@contextlib.contextmanager
def cm():
    try:
        yield
    except GeneratorExit:
        # cleanup code here
        ...
```

### Additional context
Add any other context about the feature request here.

### PR DESCRIPTION
**Pull request by rhyn0**
- add: initial layout for new error
- add: test cases for new checker
- add: first pass at new checker

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

Add a checker for `contextlib.contextmanager` decorated generator functions that don't have cleanup handling for GeneratorExit.

Closes #2832

### CODE REVIEW

File: pylint/checkers/base/function_checker.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        :param node: Node to get name of
+        :type node: nodes.NodeNG
+        :return: string name
+        :rtype: str
+        """
+        if hasattr(node, "name"):
+            return node.name  # type: ignore[no-untype-def]
+        if hasattr(node, "attrname"):
+            return node.attrname  # type: ignore[no-untype-def]
+        return node.as_string()  # type: ignore[no-untype-def]
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base/function_checker.py:126]: We have this problem approximately everywhere and generally we check the node type with ``isinstance``.

File: pylint/checkers/base/function_checker.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+        If a yield keyword is used inside a Try block, return the Try node.
+        Otherwise return the node of the yield.
+        Returns all such occurrences.
+
+        :param node: Function root to check
+        :type node: nodes.FunctionDef
+        :return: List of found nodes
+        :rtype: list[_GeneratorTree]
+        """
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/base/function_checker.py:140]: Look like it could use ``for child in node.nodes_of_class(nodes.Yield):``

**Review by Pierre-Sassoulas — COMMENTED** [review]
Thank you for your contribution, a lot of work went into this clearly. Would you mind adding a good/bad example in https://github.com/pylint-dev/pylint/tree/main/doc/data/messages too ? It's not enforced yet but we're close to being fully documented :)

File: doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+@contextlib.contextmanager
+def cm():
+    contextvar = "acquired context"
+    print("cm enter")
+    yield contextvar
+    print("cm exit")
+
+
+def genfunc_with_cm():  # [contextmanager-generator-missing-cleanup]
```
Comment (DanielNoord) [inline doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py:12]: Why are we raising the message on the function and no on the context manager itself?

File: pylint/checkers/base/function_checker.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+        )
+    }
+
+    def __init__(self, linter: PyLinter) -> None:
+        """Initialize the checker and container to store context-managers."""
+        super().__init__(linter)
+        # as this walks FunctionDef nodes, if it is wrapped in a contextmanager
+        # add it here for checking in generator functions
+        # if encounter generator FunctionDef, then check if it uses a contextmanager and conditionally add here
+        # TODO: should this string be function name or module import name (foo v mod.bar.foo)
```
Comment (DanielNoord) [inline pylint/checkers/base/function_checker.py:36]: Module import name is probably better.

**Review by DanielNoord — COMMENTED** [review]
I have a bit of trouble trying to read the code as there is so much state involved.

It feels like this will only warn whenever the context manager is defined in the same module as the module it is used in, or if the module with the manager is linted before the using module. Shouldn't we be trying to `safe_infer` the decorator and seeing if it is indeed a "faulty" context manager? That would make it robust against imports as well.
We can still consider caching the result of the check using some state, but since `safe_infer` should be cached I'm not even sure if that is needed.

**Comment by rhyn0** [comment]
Ahh I didn't know about 'safe_infer'. That would be helpful because I can find the FunctionDef of the used item in the With. 

_Note for later_
'safe_infer(with_node.items[0][0])'

That should allow me to remove the odd state objects in the init.

As for the warnings being on using generator was due to new background from the issue #2832. Using certain context-managers in a generator will cause context leaks which should be warned on the using function rather than the context-manager.

File: doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py (hunk 3 above)
Comment (jacobtylerwalls) [reply doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py:12]: Yeah, on the issue I suggested that we will want to sniff out only the use cases where it looks like a resource is being managed, AND where our prescription (add cleanup code) makes sense.

You can call a context manager either with a with statement or a decorator. If you call it with a decorator, the cleanup code in the manager won't even make a difference, so we shouldn't warn. And if you call it with a with statement, and you immediately discard the yielded value, it suggests it's not something that needs to be cleaned up.

All of that is easier to tell from the invocation of the manager than from the manager itself, even if it doesn't seem like the purest place to raise the message. Thoughts?

File: pylint/checkers/base/function_checker.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            # no try blocks at all, so checks after this line do not apply
+            return True
+        # if the contextmanager has a finally block, then it is fine
+        if all(try_node.finalbody for try_node in try_with_yield_nodes):
+            return False
+        # if the contextmanager catches GeneratorExit, then it is fine
+        if all(
+            any(
+                isinstance(handler.type, nodes.Name)
+                and handler.type.name == "GeneratorExit"
```
Comment (jacobtylerwalls) [inline pylint/checkers/base/function_checker.py:177]: Check me on this, but I think elsewhere we check the `qname()` against `builtins.GeneratorExit`, something like that, just to be extra careful in case someone has polluted their namespace 😆

File: pylint/checkers/base/function_checker.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        if any(
+            yield_node.value is None or isinstance(yield_node.value, nodes.Const)
+            for yield_node in yield_nodes
+        ):
+            return False
+        # if function body has multiple Try, filter down to the ones that have a yield node
+        try_with_yield_nodes = [
+            try_node
+            for try_node in node.nodes_of_class(nodes.Try)
+            if list(try_node.nodes_of_class(nodes.Yield))
```
Comment (jacobtylerwalls) [inline pylint/checkers/base/function_checker.py:165]: nit: probably don't need to list-ify this generator

File: doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py (hunk 3 above)
Comment (jacobtylerwalls) [reply doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py:12]: Probably worth a short explanation in details.rst and a link out to the python discussion.

File: pylint/checkers/base/function_checker.py [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+            - only if a with/async with statement is used
+            - only if the val in `with cm() as val` is not discarded, e.g. not for uses like with cm():
+
+        :param node: Node to check
+        :type node: nodes.With
+        :return: True if fails, False otherwise
+        :rtype: bool
+        """
+        # if the with/async with statement discards the value, then return False
+        if all(item[1] is None for item in node.items):
```
Comment (jacobtylerwalls) [inline pylint/checkers/base/function_checker.py:199]: I think we need to check the with statement that we care about, e.g.

```py
with (a(), b() as care_about_it):
    ....
```

a's yield is discarded, b's is not, but what if only `a` is the one with a `try`.

File: pylint/checkers/base/function_checker.py [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+            return False
+
+        return True
+
+    @staticmethod
+    def _get_name(node: nodes.NodeNG) -> str:
+        """Helper function to get the final name component of a node.
```
Comment (jacobtylerwalls) [inline pylint/checkers/base/function_checker.py:206]: Haven't looked in detail, but you might compare what you have here to ``NodeNG.repr_name()`` and let us know if you see gaps.

File: doc/data/messages/c/contextmanager-generator-missing-cleanup/good.py [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+def good_cm_yield_none():
+    print("good cm enter")
+    yield
+    print("good cm exit")
+
+
+def genfunc_with_none_yield():
+    with good_cm_yield_none() as var:
+        print(var)
+        yield "discarded"
```
[the remark this answers is not in the document]
Comment (rhyn0) [reply doc/data/messages/c/contextmanager-generator-missing-cleanup/good.py:34]: woops, that was a lot of copy paste on my end

File: doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py (hunk 3 above)
Comment (rhyn0) [reply doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py:12]: sorry for slow follow up on this. Should i include a `details.rst` for the message with some more background and reason for the message in `doc/data/`.

Think I would reference the 2 articles you link in the issue and summarizations of those problems, while also contrasting what @DanielNoord points out about generator exhaustion cases

File: pylint/checkers/base/function_checker.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+from pylint.checkers.base.basic_checker import _BasicChecker
+from pylint.lint.pylinter import PyLinter
+
+
+class FunctionChecker(_BasicChecker):
+    """Check if a function definition handles possible side effects."""
+
+    msgs = {
+        "W0135": (
+            "The code after line %s will never be executed ('GeneratorExit' needs to be handled)",
```
[the remark this answers is not in the document]
Comment (rhyn0) [reply pylint/checkers/base/function_checker.py:22]: Yah this was a pretty rough guess at what to use based on the original direction. Since we raise on contextmanager decorated function, should this instead talk about the context might not be properly cleaned up based on usage?

File: doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py (hunk 3 above)
Comment (jacobtylerwalls) [reply doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py:12]: Yes please!

File: pylint/checkers/base/function_checker.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jacobtylerwalls) [reply pylint/checkers/base/function_checker.py:22]: Sounds good, happy to review the text when you give it a go.

**Comment by Pierre-Sassoulas** [comment]
Does this need reviews or do you need to polish some things still @rhyn0 ? I would be glad to have this in 3.2.0 😄

**Comment by jacobtylerwalls** [comment]
From what I can tell it's very close:
- some unresolved documentation comments
- docs check is failing
- TODO statements about yield and asyncwith

Happy to give the whole thing a once-over when those are in 👍

File: tests/functional/c/consider/consider_using_with.py [hunk 11]
```diff
@@ -1,4 +1,5 @@
 # pylint: disable=missing-function-docstring, missing-module-docstring, invalid-name, import-outside-toplevel
+# pylint: disable=contextmanager-generator-missing-cleanup
```
Comment (jacobtylerwalls) [inline tests/functional/c/consider/consider_using_with.py:2]: I'm showing this is no longer needed.

File: doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py (hunk 3 above)
Comment (rhyn0) [reply doc/data/messages/c/contextmanager-generator-missing-cleanup/bad.py:12]: @jacobtylerwalls added the `details.rst` Let me know if you think I should add anything else there or general grammar and reading

File: pylint/checkers/base/function_checker.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (rhyn0) [reply pylint/checkers/base/function_checker.py]: @jacobtylerwalls edited the message to be more explicity about the contextmanager might not exit cleanly and how to handle

**Comment by rhyn0** [comment]
Failing ReadTheDocs error seems to be the following
```
doc/user_guide/messages/warning/implicit-str-concat.rst:55: WARNING: Lexing literal_block '[STRING_CONSTANT]\ncheck-str-concat-over-line-jumps = yes' as "toml" resulted in an error at token: 'y'. Retrying in relaxed mode.
```

Not sure if I need to change this here

File: doc/data/messages/c/contextmanager-generator-missing-cleanup/details.rst [hunk 12]
```diff
@@ -0,0 +1,10 @@
+Instantiating and using a contextmanager inside a generator function can
+result in unexpected behavior if there is an expectation that the context is only
+available for the generator function. In the case that the generator is not closed or destroyed
+then the context manager is held suspended as is.
+
+This message warns on the generator function instead of the contextmanager function
+because the ways to use a contextmanager are many.
+A contextmanager can be used as a decorator (which immediately has ``__enter__``/``__exit__`` applied)
+and the use ``as ...`` or discard of the return value also implies whether the context needs cleanup or not.
+So for this message, warning the invoker of the contextmanager is important.
```
Comment (jacobtylerwalls) [inline doc/data/messages/c/contextmanager-generator-missing-cleanup/details.rst:10]: Thanks, this is really great. A really subtle set of circumstances explained very clearly.

File: doc/data/messages/c/contextmanager-generator-missing-cleanup/details.rst [hunk 13]
```diff
@@ -0,0 +1,10 @@
+Instantiating and using a contextmanager inside a generator function can
+result in unexpected behavior if there is an expectation that the context is only
+available for the generator function. In the case that the generator is not closed or destroyed
+then the context manager is held suspended as is.
+
+This message warns on the generator function instead of the contextmanager function
+because the ways to use a contextmanager are many.
+A contextmanager can be used as a decorator (which immediately has ``__enter__``/``__exit__`` applied)
+and the use ``as ...`` or discard of the return value also implies whether the context needs cleanup or not.
```
Comment (jacobtylerwalls) [inline doc/data/messages/c/contextmanager-generator-missing-cleanup/details.rst:9]: use -> use of

File: doc/whatsnew/fragments/2832.feature [hunk 14]
```diff
@@ -0,0 +1,6 @@
+Checks for generators that use contextmanagers that don't handle cleanup properly.
+Is meant to raise visibilty on the case that a generator is not fully exhausted and the contextmanager is not cleaned up properly.
+A contextmanager must yield a non constant value and not handle cleanup for GeneratorExit.
```
Comment (jacobtylerwalls) [inline doc/whatsnew/fragments/2832.feature:3]:
```suggestion
A contextmanager must yield a non-constant value and not handle cleanup for GeneratorExit.
```

File: doc/whatsnew/fragments/2832.feature [hunk 15]
```diff
@@ -0,0 +1,6 @@
+Checks for generators that use contextmanagers that don't handle cleanup properly.
```
Comment (jacobtylerwalls) [inline doc/whatsnew/fragments/2832.feature:1]: (FYI, if you use the `new_check` suffix instead of `feature` it will be grouped a little bit more helpfully.)

File: pylint/checkers/base/function_checker.py (hunk 6 above)
Comment (jacobtylerwalls) [reply pylint/checkers/base/function_checker.py:121]: Maybe I'm missing something obvious, but it seemed on a first impression that `list` is more expensive than `any` in this instance. Or is that not so?

File: tests/functional/c/consider/consider_using_with_open.py [hunk 16]
```diff
@@ -52,17 +55,22 @@ def test_open_inside_with_block():
...
+    with (
+        open(file1, encoding="utf-8") if which else open(file2, encoding="utf-8")
+    ) as input_file:  # must not trigger
         return input_file.read()
 
 
 def test_single_line_with(file1):
-    with open(file1, encoding="utf-8"): return file1.read()  # must not trigger
+    with open(file1, encoding="utf-8"):
+        return file1.read()  # must not trigger
```
Comment (jacobtylerwalls) [inline tests/functional/c/consider/consider_using_with_open.py:66]: Would you mind reverting the unrelated formatting changes? As you can tell from the test method names, they're testing edge cases relating to oddly formatted code. We don't want to unintentionally degrade test coverage.

**Review by jacobtylerwalls — COMMENTED** [review]
Looking forward to merging this shortly. Just some minor feedback.

**Comment by jacobtylerwalls** [comment]
> Not sure if I need to change this here

Good call, let's not block your work on that. If it hasn't been fixed on main yet, please open an issue!

**Review by jacobtylerwalls — APPROVED** [review]
Thanks for the new check @rhyn0!
